src/WebJobs.Script.Grpc/Server/ExtensionsCompositeEndpointDataSource.cs

```diff
@@ -0,0 +1,300 @@
+                {
+                    disposable.Dispose();
+                }
+            }
+
+            _cts?.Dispose();
+        }
+
+        private Endpoint WrapEndpoint(Endpoint endpoint)
+        {
+            static RequestDelegate CreateDelegate(RequestDelegate next, IServiceProvider services)
+            {
+                // Incoming HttpContext has the gRPC script host services. Create a scope
+                // for the JobHost services and swap out the contexts request services.
+                return async context =>
+                {
+                    if (next is null)
+                    {
+                        return;
+                    }
+
+                    IServiceProvider original = context.RequestServices;
+
+                    try
+                    {
+                        await using AsyncServiceScope scope = services.CreateAsyncScope();
+                        context.RequestServices = scope.ServiceProvider;
+                        await next(context);
+                    }
+                    finally
+                    {
+                        context.RequestServices = original;
+                    }
+                };
+            }
+
+            if (endpoint is not RouteEndpoint route)
+            {
+                // We only wrap URL-routeable endpoints (ie: RouteEndpoint).
+                return endpoint;
+            }
+
+            IServiceProvider services = _extensionServices
+                ?? throw new InvalidOperationException(
+                    "Trying to register extension endpoints, but no extension IServiceProvider available.");
+
+            return new RouteEndpoint(
+                CreateDelegate(route.RequestDelegate, services),
+                route.RoutePattern,
+                route.Order,
+                route.Metadata,
+                route.DisplayName);
+        }
+
+        [MemberNotNull(nameof(_consumerChangeToken))]
+        private void EnsureChangeTokenInitialized()
+        {
+            if (_consumerChangeToken is not null)
+            {
+                return;
+            }
+
+            lock (_lock)
+            {
+                if (_consumerChangeToken is not null)
+                {
+                    return;
+                }
+
+                // This is our first time initializing the change token, so the collection has "changed" from nothing.
+                CreateChangeTokenUnsynchronized(collectionChanged: true);
+            }
+        }
+
+        [MemberNotNull(nameof(_consumerChangeToken))]
+        private void CreateChangeTokenUnsynchronized(bool collectionChanged)
+        {
+            CancellationTokenSource cts = new();
+
+            if (collectionChanged)
+            {
+                _changeTokenRegistrations = new();
+                foreach (var dataSource in _dataSources)
+                {
+                    _changeTokenRegistrations.Add(ChangeToken.OnChange(
```

Comment:
I'm not super familiar with this, but was wondering if using [CompositeChangeToken](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/change-tokens?view=aspnetcore-7.0#compositechangetoken-class) works here instead of manually managing a list of change token registrations? 

If this doesn't work for us, I'd love to learn more about why (just for my learning)

Reply:
I do that that could be used here with a bit more effort. However, this class is a modified copy of [CompositeEndpointDataSource](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/CompositeEndpointDataSource.cs) from AspNetCore, and this change token pattern comes from there. Given this code is already proven by AspNetCore, I would rather not alter it more than I have to.